proposals/nnnn-trailing-comma-lists.md

```diff
@@ -0,0 +1,232 @@
+
+## Proposed solution
+
+This proposal adds support for trailing commas in comma-separated lists when there's a clear terminator, which are the following:
+
+- Tuples and tuple patterns 
+
+```swift
+(1, 2,)
+let block: (Int, Int,) -> Void = { (a, b,) in  }
+let (a, b,) = (1, 2,)
+for (a, b,) in zip(s1, s2) { }
+```
+
+- Declaration and call site for arguments lists of initializers, functions, associated values, built-in attributes, expression macros, attached macros and availability spec.
+
+```swift
+
+func foo(a: Int, b: Int,) { }
+
+foo(a: 1, b: 1,)
+
+struct S {
+    init(a: Int, b: Int,) { }
+}
+
+enum E {
+    case foo(a: Int, b: Int,)
+}
+
+@Foo(1, 2, 3,) 
+struct S { }
+
+f(_: @foo(1, 2,) Int)
+
+#foo(1, 2,)
+
+struct S {
+    #foo(1, 2,)
+}
+
+if #unavailable(iOS 15, watchOS 9,) { }
+
+```
+- Subscripts, including key path subscripts.
+
+```swift
+let value = m[x, y,]
+
+let keyPath = \Foo.bar[x,y,]  
+
+f(\.[x,y,])
+```
+
+- `if`, `guard` and `while` condition lists.
+
+```swift
+if a, b, { }
+while a, b, { }
+guard a, b, else { }
+```
+
+- `switch` case labels.
+
+```swift
+switch number {
+    case 1, 2,:
+        ...
+    default:
+        ..
+}
+```
+
+- Closure capture lists.
+
+```swift
+{ [a, b,] in }
+```
+
+- Inheritance clauses.
+
+```swift
+struct S: P1, P2, P3, { }
+```
+
+- Generic parameters.
+
+```swift
+struct S<T1, T2, T3,> { }
+```
+
+- Generic `where` clauses.
+
+```swift
+struct S<T1, T2, T3> where T1: P1, T2: P2, { }
+```
+
+- String interpolation
+
+```swift
+let s = "\(1, 2,)"
+```
+
+## Detailed Design
+
+Trailing comma will be supported in comma-separated lists whenever there is a terminator clear enough that the parser can determine the end of the list. The terminator can be the symbols like `)`, `]`, `>`, `{` and `:`, a keyword like `where` or a pattern code like the body of a `if`, `guard` and `while` statement.
+
+Note that the requirement for a terminator means that the following cases will not support trailing comma:
+
+Enum case label lists:
+
+```swift
+enum E {
+  case a, b, c, // ❌ Expected identifier after comma in enum 'case' declaration
+}
+```
+
+Inheritance clauses for associated type in a protocol declaration:
+
+```swift
+protocol Foo {
+  associatedtype T: P1, P2, // ❌ Expected type
+}
+```
+
+Generic `where` clauses for initializers and functions in protocol declaration:
+
+```swift
+protocol Foo {
+  func f<T1, T2>(a: T1, b: T2) where T1: P1, T2: P2, // ❌ Expected type
+}
+```
+
+Trailing comma will be allowed in single-element lists but not in zero-element lists, since the trailing comma is actually attached to the last element. Supporting a zero-element list would require supporting _leading_ commas, which isn't what this proposal is about.
```

Comment:
```suggestion
Trailing commas will be allowed in single-element lists but not in zero-element lists, since the trailing comma is actually attached to the last element. Supporting a zero-element list would require supporting _leading_ commas, which isn't what this proposal is about.
```